intermediate_source/model_parallel_tutorial.py

```diff
@@ -0,0 +1,321 @@
+# but we can get a reasonable understanding of the execution times.
+
+
+import torchvision.models as models
+import timeit
+
+num_batches = 3
+batch_size = 120
+image_w = 128
+image_h = 128
+
+
+def train(model):
+    model.train(True)
+    loss_fn = nn.MSELoss()
+    optimizer = optim.SGD(model.parameters(), lr=0.001)
+
+    one_hot_indices = torch.LongTensor(batch_size) \
+                           .random_(0, num_classes) \
+                           .view(batch_size, 1)
+
+    for _ in range(num_batches):
+        # generate random inputs and labels
+        inputs = torch.randn(batch_size, 3, image_w, image_h)
+        labels = torch.zeros(batch_size, num_classes) \
+                      .scatter_(1, one_hot_indices, 1)
+
+        # run forward pass
+        optimizer.zero_grad()
+        outputs = model(inputs.cuda())
+
+        # run backward pass
+        labels = labels.to(outputs.device)
+        loss_fn(outputs, labels).backward()
+        optimizer.step()
+
+
+######################################################################
+# The ``train(model)`` method above uses ``nn.MSELoss`` as the loss function,
+# and ``optim.SGD`` as the optimizer. It mimics training on ``128 X 128``
+# images which are organized into 3 batches where each batch contains 120
+# images. Then, we use ``timeit`` to run the ``train(model)`` method 10 times
+# and plot the execution times with standard deviations.
+
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+num_repeat = 10
+
+stmt = "train(model)"
+
+setup = "from __main__ import train, ModelParallelResNet50;" + \
+        "model = ModelParallelResNet50()"
+mp_run_times = timeit.repeat(stmt, setup, number=1, repeat=num_repeat)
+mp_mean, mp_std = np.mean(mp_run_times), np.std(mp_run_times)
+
+setup = "from __main__ import train, num_classes;" + \
+        "import torchvision.models as models;" + \
+        "model = models.resnet50(num_classes=num_classes).cuda()"
+rn_run_times = timeit.repeat(stmt, setup, number=1, repeat=num_repeat)
+rn_mean, rn_std = np.mean(rn_run_times), np.std(rn_run_times)
+
+
+def plot(means, stds, labels, fig_name):
+    fig, ax = plt.subplots()
+    ax.bar(np.arange(len(means)), means, yerr=stds,
+           align='center', alpha=0.5, ecolor='red', capsize=10, width=0.6)
+    ax.set_ylabel('ResNet50 Execution Time (Second)')
+    ax.set_xticks(np.arange(len(means)))
+    ax.set_xticklabels(labels)
+    ax.yaxis.grid(True)
+    plt.tight_layout()
+    plt.savefig(fig_name)
+
+
+plot([mp_mean, rn_mean],
+     [mp_std, rn_std],
+     ['Model Parallel', 'Single GPU'],
+     'mp_vs_rn.png')
+
+
+######################################################################
+#
+# .. figure:: /_static/img/model-parallel-images/mp_vs_rn.png
+#    :alt:
+#
+# The result shows that the execution time of model parallel implementation is
+# ``4.02/3.75-1=7%`` longer than the existing single-GPU implementation. There
```

Comment:
you might write something like, "so we can conclude there is roughly 7% overhead in copying tensors back and forth across the GPUs" (or similar).